### PR TITLE
Firecracker exits before returning API response FIX

### PR DESCRIPTION
**Pull request by berciuliviu**
## Reason for This PR

This PR addresses the following [bug](https://github.com/firecracker-microvm/firecracker/issues/2116). For now, the status is RFC since integration tests have to be added.

## Description of Changes

The important points of this change are the following:

- the VCPU errors that killed the process before, are now allowed to return to the API
thread.
- the API thread decides if an error is considered **fatal** or not. New errors can be added in the future which are deemed fatal
- in case of a **fatal** error, the Firecracker process is killed by the API thread,
since the VMM thread is already compromised
- this mechanism abides by the rules of the [Snapshotting specification](https://github.com/firecracker-microvm/firecracker/blob/master/docs/snapshotting/snapshot-support.md) which says that,
in the case of an error when _loading a snapshot_, _"a specific error is reported and then the current Firecracker process is ended"_

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [ ] All commits in this PR are signed (`git commit -s`).
- [ ] The reason for this PR is clearly provided (issue no. or explanation).
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] Any newly added `unsafe` code is properly documented.
- [ ] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

### CODE REVIEW

File: src/api_server/src/lib.rs [hunk 1]
```diff
@@ -217,6 +231,19 @@ impl ApiServer {
         response
     }
 
+    fn check_for_fatal_error(&self, response: &std::result::Result<VmmData, VmmActionError>) {
```
Comment (acatangiu) [inline src/api_server/src/lib.rs:234]: If you make `&self` mutable, you can avoid the need for using `RefCell`.

```suggestion
    fn check_for_fatal_error(&mut self, response: &std::result::Result<VmmData, VmmActionError>) {
```

File: src/api_server/src/lib.rs [hunk 2]
```diff
@@ -67,6 +70,9 @@ pub struct ApiServer {
     /// FD on which we notify the VMM that we have sent at least one
     /// `VmmRequest`.
     to_vmm_fd: EventFd,
+    /// If this flag is set with some value, the process encountered
+    /// a fatal error and it is going to exit.
+    encountered_fatal_error: RefCell<Option<u8>>,
```
Comment (acatangiu) [inline src/api_server/src/lib.rs:75]: You should save the error so you can log it before exiting. Also, ideally avoid `RefCell` unless necessary.

```suggestion
    fatal_error_encountered: Option<VmmActionError>,
```

File: src/api_server/src/lib.rs [hunk 3]
```diff
@@ -129,6 +136,12 @@ impl ApiServer {
         loop {
             match server.requests() {
                 Ok(request_vec) => {
+                    if let Some(exit_code) = *self.encountered_fatal_error.borrow() {
```
Comment (acatangiu) [inline src/api_server/src/lib.rs:139]: This needs to happen inside the loop below; right after sending the HTTP response (after `server.respond()`).

Ideally it should also flush the http socket to make sure the response is sent before exiting the process.

File: src/api_server/src/lib.rs [hunk 4]
```diff
@@ -129,6 +136,12 @@ impl ApiServer {
...
+                        unsafe {
+                            error!("Fatal error with exit code: {}", exit_code);
```
Comment (acatangiu) [inline src/api_server/src/lib.rs:141]: nit: take this `error!` out of the `unsafe` block.

File: src/api_server/src/lib.rs [hunk 5]
```diff
@@ -217,6 +231,19 @@ impl ApiServer {
         response
     }
 
+    fn check_for_fatal_error(&self, response: &std::result::Result<VmmData, VmmActionError>) {
+        if let Err(e) = response {
+            let mut is_fatal = self.encountered_fatal_error.borrow_mut();
+            // Errors considered as fatal are added to this 'match'
+            match e {
```
Comment (acatangiu) [inline src/api_server/src/lib.rs:238]: save this `e` to log it before exiting.

File: src/api_server/src/lib.rs [hunk 6]
```diff
@@ -67,6 +70,9 @@ pub struct ApiServer {
     /// FD on which we notify the VMM that we have sent at least one
     /// `VmmRequest`.
     to_vmm_fd: EventFd,
+    /// If this flag is set with some value, the process encountered
+    /// a fatal error and it is going to exit.
```
Comment (acatangiu) [inline src/api_server/src/lib.rs:74]: nit:

```suggestion
    /// a fatal error and it is going to exit once it sends any pending API response.
```

**Review by acatangiu — COMMENTED** [review]
Love the approach, only have punctual code comments.

File: src/api_server/src/lib.rs (hunk 5 above)
Comment (berciuliviu) [reply src/api_server/src/lib.rs:238]: What do you think if I create a separate structure containing both the error and an exit code meaningful to it? This way I could log both the error and pass the exit code to libc::exit.

File: src/api_server/src/lib.rs (hunk 1 above)
Comment (berciuliviu) [reply src/api_server/src/lib.rs:234]: The problem here is the following: `check_for_fatal_error` is called by `serve_vmm_action_request`, which is called by `handle_request`. Both these functions don't have a _mutable self_ in their function header. If I am to add `&mut self` to both functions, it still wouldn't work since, on line **152**, `handle_request` is called from a closure, for which Rust says "cannot borrow `*self` as mutable, as `Fn` closures cannot mutate their captured variables". This is also the reason I tried using a RefCell. An Arc<Mutex<>> also worked, but I thought it would be too resourceful for a variable that is not used across threads and contains relatively small data.

File: src/api_server/src/lib.rs (hunk 4 above)
Comment (berciuliviu) [reply src/api_server/src/lib.rs:141]: Will do.

File: src/api_server/src/lib.rs (hunk 3 above)
Comment (berciuliviu) [reply src/api_server/src/lib.rs:139]: While I agree to this approach, the function `server.requests()` also does the job of responding to the client when there are `epoll::EventSet::OUT` events to be processed. Do you think it would be feasible to call again `server.requests()` before exiting the process, so that it could flush the current response queue?

File: src/api_server/src/lib.rs (hunk 2 above)
Comment (berciuliviu) [reply src/api_server/src/lib.rs:75]: Will do. The reason for RefCell usage is explained on a comment below.

File: src/api_server/src/lib.rs (hunk 1 above)
Comment (berciuliviu) [reply src/api_server/src/lib.rs:234]: After discussing with @acatangiu , we have decided to modify the `ServerRequest::process` function to allow for a mutable callable, instead of an immutable one. This change allows us the change the chain of functions above to accept a `&mut self`, allowing us to change the _RefCell_ to a _boolean_.

File: src/api_server/src/lib.rs (hunk 3 above)
Comment (berciuliviu) [reply src/api_server/src/lib.rs:139]: After discussing with @acatangiu , who guided me to reach out to @georgepisaltu to dive deeper into the case, we have decided on the following: the verification for encountering fatal errors will be done after the server response at line 148. If we have a fatal error, a new HttpServer function will check if we have pending outgoing writes in a loop. If we do, we call `server.requests()` to address the current response queue, while also returning a 500 Internal Server Error to any new incoming requests. We kill the process afterwards.

File: src/api_server/src/lib.rs [hunk 7]
```diff
@@ -217,6 +240,15 @@ impl ApiServer {
         response
     }
 
+    fn check_for_fatal_error(&mut self, response: &std::result::Result<VmmData, VmmActionError>) {
+        if let Err(e) = response {
+            // Errors considered as fatal are added here
+            if let VmmActionError::LoadSnapshot(_) = e {
```
Comment (acatangiu) [inline src/api_server/src/lib.rs:246]: nit: can be shortened as:
```suggestion
        // Errors considered as fatal are added here
        if let Err(VmmActionError::LoadSnapshot(_)) = response {
```

File: src/micro_http/src/server.rs [hunk 8]
```diff
@@ -379,6 +379,19 @@ impl HttpServer {
...
     }
 
+    /// This function is responsible with flushing any remaining outgoing
+    /// requests on the server.
+    ///
+    /// Note that this function can block the thread on write, since the
+    /// operation is blocking.
+    pub fn flush_outgoing_writes(&mut self) {
+        for (_, connection) in self.connections.iter_mut() {
+            if connection.state == ClientConnectionState::AwaitingOutgoing {
```
Comment (acatangiu) [inline src/micro_http/src/server.rs:389]: should this be a loop?
```suggestion
            while connection.state == ClientConnectionState::AwaitingOutgoing {
```

**Review by acatangiu — COMMENTED** [review]
Looks good!

File: src/micro_http/src/server.rs (hunk 8 above)
Comment (sandreim) [reply src/micro_http/src/server.rs:389]: Can it be an infinite loop ? What if we cannot send data back to the client ?

File: src/api_server/src/lib.rs [hunk 9]
```diff
@@ -67,6 +69,9 @@ pub struct ApiServer {
     /// FD on which we notify the VMM that we have sent at least one
     /// `VmmRequest`.
     to_vmm_fd: EventFd,
+    /// If this flag is set, the process encountered  a fatal error
+    /// and it is going to exit once it sends any pending API response.
+    encountered_fatal_error: bool,
```
Comment (sandreim) [inline src/api_server/src/lib.rs:74]: I would call it vmm_fatal_error.

File: src/micro_http/src/server.rs [hunk 10]
```diff
@@ -379,6 +379,19 @@ impl HttpServer {
         Ok(parsed_requests)
     }
 
+    /// This function is responsible with flushing any remaining outgoing
+    /// requests on the server.
+    ///
+    /// Note that this function can block the thread on write, since the
+    /// operation is blocking.
+    pub fn flush_outgoing_writes(&mut self) {
```
Comment (sandreim) [inline src/micro_http/src/server.rs:387]: I think we should also call close() on the fds after writing all data.

**Comment by sandreim** [comment]
The non-draft PR should also include the changelog update on the behavior change.

File: src/micro_http/src/server.rs (hunk 8 above)
Comment (berciuliviu) [reply src/micro_http/src/server.rs:389]: @acatangiu yes, this could also be a loop. Thumbs up for the suggestion!
@sandreim yes, it can block on write if the operation waits on the client

Comment (sandreim) [reply src/micro_http/src/server.rs:389]: Can we have a timeout for the blocking case ? so that the behavior is predictible.

File: src/micro_http/src/server.rs (hunk 10 above)
Comment (acatangiu) [reply src/micro_http/src/server.rs:387]: is it necessary? doesn't the same thing happen (process buffers are flushed) during/because-of `libc::exit()`?

File: src/micro_http/src/server.rs (hunk 8 above)
Comment (acatangiu) [reply src/micro_http/src/server.rs:389]: hmm.. I can't think of any non-complex solution to handling the blocking case.

Is the extra complexity necessary though? Isn't it the same behavior as all other API responses?

It is a good catch @sandreim, but **if there is no easy way to do it**, I say take it as is.
I wouldn't invest more time and adding more complexity for these unlikely error paths which are irrelevant in production scenarios, since that comes at the expense of investing in things with much higher impact and value.

File: src/api_server/src/lib.rs (hunk 9 above)
Comment (berciuliviu) [reply src/api_server/src/lib.rs:74]: Will do

File: src/api_server/src/lib.rs (hunk 7 above)
Comment (berciuliviu) [reply src/api_server/src/lib.rs:246]: Will do

File: src/micro_http/src/server.rs (hunk 10 above)
Comment (berciuliviu) [reply src/micro_http/src/server.rs:387]: I agree with @acatangiu

File: src/micro_http/src/server.rs [hunk 11]
```diff
@@ -379,6 +379,21 @@ impl HttpServer {
...
+        for (_, connection) in self.connections.iter_mut() {
+            while connection.state == ClientConnectionState::AwaitingOutgoing {
+                if connection.write().is_err() {
```
Comment (berciuliviu) [inline src/micro_http/src/server.rs:390]: I have added this error handling for the case in which an error such as a ConnectionClosed would not allow the function to set the proper ClientConnectionState. In that case, we want to break out of the loop.

File: src/api_server/src/lib.rs [hunk 12]
```diff
@@ -67,6 +69,9 @@ pub struct ApiServer {
     /// FD on which we notify the VMM that we have sent at least one
     /// `VmmRequest`.
     to_vmm_fd: EventFd,
+    /// If this flag is set, the process encountered  a fatal error
```
Comment (acatangiu) [inline src/api_server/src/lib.rs:72]: nit:
```suggestion
    /// If this flag is set, the process encountered a fatal error
```

File: tests/integration_tests/functional/test_snapshot_basic.py [hunk 13]
```diff
@@ -351,6 +352,82 @@ def test_5_inc_snapshots(network_config,
...
+                                       ssh_key,
+                                       SnapshotType.FULL)
+
+    vm.kill()
+
+    # Corrupt the memory file of the snapshot
+    logger.info("Corrupt the memory file %s", snapshot.mem)
+
+    mem_fd = open(snapshot.mem, "wb")
+    mem_fd.write(bytearray([0xde, 0xad, 0xbe, 0xef, 0x00]))
```
Comment (acatangiu) [inline tests/integration_tests/functional/test_snapshot_basic.py:399]: How does this guarantee the `LoadSnapshot` API command will fail? Wouldn't it be more robust to just use an empty or completely zeroed out snapshot vm-state file?

Memory corruptions will affect guest and while we do have _some_ validations for it, it is not guaranteed that the snapshot load operation will fail because of it.

File: tests/integration_tests/functional/test_snapshot_basic.py [hunk 14]
```diff
@@ -351,6 +352,82 @@ def test_5_inc_snapshots(network_config,
...
+
+    response = vm.snapshot_load.put(mem_file_path=jailed_mem,
+                                    snapshot_path=jailed_vmstate)
+
+    logger.info("Response status code %d, content: %s.", response.status_code, response.text)
+    assert vm.api_session.is_status_bad_request(response.status_code)
+
+    logger.info("The Firecracker PID is %d.", vm.jailer_clone_pid)
+    try:
+        os.kill(vm.jailer_clone_pid, 0)
```
Comment (acatangiu) [inline tests/integration_tests/functional/test_snapshot_basic.py:425]: This is brittle, the pid could be reused, chance is small for that to happen, but I would avoid it if possible.

`os.kill()` can also inadvertently kill other processes that might reuse the pid.

Instead of using `os.kill` you could check process age. See this for example https://unix.stackexchange.com/questions/238641/get-process-age-from-command-line

We can get `jailer_pid` process `start_time` using `/proc/<pid>/stat`, before calling the API. Then recheck after the API call, and it should be either:
 - no process with that PID 👍 
 or
 - process with that PID has newer `start_time` 👍 

Anything else: 👎

File: tests/integration_tests/functional/test_snapshot_basic.py (hunk 13 above)
Comment (berciuliviu) [reply tests/integration_tests/functional/test_snapshot_basic.py:399]: You are correct. I will empty the file in this case.

File: tests/integration_tests/functional/test_snapshot_basic.py (hunk 14 above)
Comment (berciuliviu) [reply tests/integration_tests/functional/test_snapshot_basic.py:425]: Understood, will do as such

Comment (berciuliviu) [reply tests/integration_tests/functional/test_snapshot_basic.py]: Done!

File: tests/integration_tests/functional/test_snapshot_basic.py [hunk 15]
```diff
@@ -352,6 +353,106 @@ def test_5_inc_snapshots(network_config,
...
 
+DEFAULT_HOST_IP = "192.168.0.1"
+DEFAULT_GUEST_IP = "192.168.0.2"
+DEFAULT_TAP_NAME = "tap0"
+DEFAULT_DEV_NAME = "eth0"
+DEFAULT_NETMASK = 30
+
+
+def _get_process_start_time(p_pid):
+    # Helper function - returns the start_time and returncode for
```
Comment (acatangiu) [inline tests/integration_tests/functional/test_snapshot_basic.py:364]: This comment is misleading:

return code provided by this function is not `return code of provided PID`.

It is actually `return code of 'stat -c%X <pid>'`

File: tests/integration_tests/functional/test_snapshot_basic.py [hunk 16]
```diff
@@ -352,6 +353,106 @@ def test_5_inc_snapshots(network_config,
...
+    mem_fd.truncate(0)
+    mem_fd.close()
+
+    # Create a new VM instance
+    vm = MicrovmBuilder(bin_cloner_path).create_basevm()
+    vm.spawn(log_level='Info')
+
+    # Hardlink all the snapshot files into the microvm jail.
+    jailed_mem = vm.create_jailed_resource(snapshot.mem)
+    jailed_vmstate = vm.create_jailed_resource(snapshot.vmstate)
```
Comment (acatangiu) [inline tests/integration_tests/functional/test_snapshot_basic.py:425]: Can you achieve the same without the first half of this test?

Instead of actually starting a base vm, creating a snapshot and corrupting that snapshot - just try to load empty `jailed_mem` and empty `jailed_vmstate`.

That would significantly cut both complexity and duration of this test.

File: tests/integration_tests/functional/test_snapshot_basic.py [hunk 17]
```diff
@@ -352,6 +353,106 @@ def test_5_inc_snapshots(network_config,
...
+                response.text)
+    assert vm.api_session.is_status_bad_request(response.status_code)
+
+    fc_exit_time, err = _get_process_start_time(vm.jailer_clone_pid)
+    if err != 0:
+        assert True
+    else:
+        if fc_time == fc_exit_time:
+            assert False, "Firecracker process is not dead!"
+        assert True
```
Comment (acatangiu) [inline tests/integration_tests/functional/test_snapshot_basic.py:453]:
```suggestion
    assert err != 0 or fc_time != fc_exit_time, "Firecracker process should have ended with error!"
```

File: tests/integration_tests/functional/test_snapshot_basic.py (hunk 15 above)
Comment (berciuliviu) [reply tests/integration_tests/functional/test_snapshot_basic.py]: Done

File: tests/integration_tests/functional/test_snapshot_basic.py (hunk 16 above)
Comment (berciuliviu) [reply tests/integration_tests/functional/test_snapshot_basic.py]: Done

File: tests/integration_tests/functional/test_snapshot_basic.py (hunk 17 above)
Comment (berciuliviu) [reply tests/integration_tests/functional/test_snapshot_basic.py:433]: Done

File: CHANGELOG.md [hunk 18]
```diff
@@ -10,6 +10,7 @@
 - Added devtool test `-m|--cpuset-mems` flag for memory confinement when tests
   run.
 - Added the virtio traditional memory ballooning device.
+- Added a mechanism for handling fatal errors, which will trigger a process kill when encountered.
```
Comment (sandreim) [inline CHANGELOG.md:13]: The process is not killed. I would make this more specific - Added a mechanism to handle vCPU/VMM errors that result in process termination.

File: src/micro_http/src/server.rs (hunk 10 above)
Comment (sandreim) [reply src/micro_http/src/server.rs:387]: I agree this also works, but as a best practice we could handle cleanup ourselves.

File: src/micro_http/src/server.rs (hunk 11 above)
Comment (sandreim) [reply src/micro_http/src/server.rs:390]: Please also log a warning/error here.

File: tests/README.md [hunk 19]
```diff
@@ -20,6 +20,12 @@ To run tests from specific directories and/or files:
 tools/devtool test -- <test_dir_or_file_path>...
 ```
 
+To run a single specific test from a file:
```
Comment (sandreim) [inline tests/README.md:23]: Please move these changes in a separate commit.

File: tests/integration_tests/functional/test_snapshot_basic.py [hunk 20]
```diff
@@ -352,6 +355,90 @@ def test_5_inc_snapshots(network_config,
...
+                            encoding='ascii')
+    out = proc.communicate()[0].strip('\n')
+    return out, proc.returncode
+
+
+@pytest.mark.skipif(
+    platform.machine() != "x86_64",
+    reason="Not supported yet."
+)
+def test_load_snapshot_failure_handling(bin_cloner_path):
```
Comment (sandreim) [inline tests/integration_tests/functional/test_snapshot_basic.py:381]: We can simplify this:

```
from framework.microvms import VMNano
def test_load_snapshot_failure_handling(bin_cloner_path):
    vm_instance = VMNano.spawn(bin_cloner_path)
    vm = vm_intsance.vm
```

File: tests/integration_tests/functional/test_snapshot_basic.py [hunk 21]
```diff
@@ -352,6 +355,90 @@ def test_5_inc_snapshots(network_config,
...
+
+    logger.info("Response status code %d, content: %s.",
+                response.status_code,
+                response.text)
+    assert vm.api_session.is_status_bad_request(response.status_code)
+
+    # Check if FC process is closed
+    seconds = 0.1
+    fc_exit_time, err = _get_process_start_time(vm.jailer_clone_pid)
+    while seconds < 300:
```
Comment (sandreim) [inline tests/integration_tests/functional/test_snapshot_basic.py:432]: You can just loop forever here, framework will timeout the test after 300 seconds.

File: src/micro_http/src/server.rs (hunk 8 above)
Comment (georgepisaltu) [reply src/micro_http/src/server.rs]: Every FD backing a stream is set to non-blocking before creating the connection (`server.rs:489`). Writes can't be blocking as we would immediately receive `EAGAIN` or `EWOULDBLOCK`. As long as we check for these two error cases, the loop can't be infinite.

File: src/api_server/src/lib.rs (hunk 12 above)
Comment (acatangiu) [reply src/api_server/src/lib.rs:72]: nit: leftover whitespace still there

File: src/api_server/src/lib.rs [hunk 22]
```diff
@@ -217,6 +241,14 @@ impl ApiServer {
         response
     }
 
+    #[cfg(target_arch = "x86_64")]
```
Comment (acatangiu) [inline src/api_server/src/lib.rs:244]: nit: instead of making the function `x86_64` specific, maybe we can make just it's internal logic specific. The function could then be called on all platforms.

File: tests/integration_tests/functional/test_snapshot_basic.py [hunk 23]
```diff
@@ -352,6 +355,90 @@ def test_5_inc_snapshots(network_config,
...
+                                    snapshot_path=jailed_vmstate)
+
+    logger.info("Response status code %d, content: %s.",
+                response.status_code,
+                response.text)
+    assert vm.api_session.is_status_bad_request(response.status_code)
+
+    # Check if FC process is closed
+    seconds = 0.1
+    fc_exit_time, err = _get_process_start_time(vm.jailer_clone_pid)
```
Comment (acatangiu) [inline tests/integration_tests/functional/test_snapshot_basic.py:431]: nit:
```suggestion
    pid_time, ret = _get_process_start_time(vm.jailer_clone_pid)
```

future `assert err != 0` makes me think we want `err` when in fact we want `ret == 0`
same for check `pid_time != fc_time` somewhat clearer than `fc_time != fc_exit_time`

File: tests/integration_tests/functional/test_snapshot_basic.py [hunk 24]
```diff
@@ -352,6 +355,90 @@ def test_5_inc_snapshots(network_config,
...
+    while seconds < 300:
+        if err != 0 or fc_time != fc_exit_time:
+            break
+        seconds += 0.1
+        time.sleep(0.1)
+        fc_exit_time, err = _get_process_start_time(vm.jailer_clone_pid)
+    assert err != 0 or fc_time != fc_exit_time, "Firecracker process should \
```
Comment (acatangiu) [inline tests/integration_tests/functional/test_snapshot_basic.py:438]: nit:
leave a helpful comment for future generations
```suggestion
    # Check that either:
    #  - pid no longer exists, or
    #  - pid exists but is a different process (different process start time)
    assert err != 0 or pid_time != fc_time, "Firecracker process should \
```

**Review by acatangiu — COMMENTED** [review]
Looks good, just a couple of nits from me and some good comments from Andrei.

File: src/micro_http/src/server.rs (hunk 11 above)
Comment (berciuliviu) [reply src/micro_http/src/server.rs]: Done

File: tests/integration_tests/functional/test_snapshot_basic.py (hunk 24 above)
Comment (berciuliviu) [reply tests/integration_tests/functional/test_snapshot_basic.py]: Done

File: src/api_server/src/lib.rs (hunk 22 above)
Comment (berciuliviu) [reply src/api_server/src/lib.rs:244]: Due to Clippy errors, this will be done better when we have more fatal errors to handle, through a larger 'match'.

File: src/micro_http/src/server.rs (hunk 8 above)
Comment (berciuliviu) [reply src/micro_http/src/server.rs]: `connection.write()`  checks for these cases through a more general StreamError approach, so we are covered in this regard. The respective does close the connection whenever it encounters an unexpected StreamError or a ConnectionClosed error.

**Review by acatangiu — DISMISSED** [review]

File: src/micro_http/src/server.rs [hunk 25]
```diff
@@ -379,6 +380,22 @@ impl HttpServer {
...
     }
 
+    /// This function is responsible with flushing any remaining outgoing
+    /// requests on the server.
+    ///
+    /// Note that this function can block the thread on write, since the
+    /// operation is blocking.
+    pub fn flush_outgoing_writes(&mut self) {
+        for (_, connection) in self.connections.iter_mut() {
+            while connection.state == ClientConnectionState::AwaitingOutgoing {
```
Comment (georgepisaltu) [inline src/micro_http/src/server.rs:390]: After the while loop, I would also set the connection state to `AwaitingIncoming`, so that if we use this method in the future in a context where we won't immediately exit the process, the state of the connection is still valid.

File: src/micro_http/src/server.rs [hunk 26]
```diff
@@ -379,6 +380,22 @@ impl HttpServer {
...
+                if let Err(e) = connection.write() {
+                    error!("Connection write error: {}", e);
```
Comment (georgepisaltu) [inline src/micro_http/src/server.rs:392]: I suggest avoiding the logging with `error!` if the error we receive from `connection.write()` is `Err(ServerError::ConnectionError(ConnectionError::InvalidWrite))`, since that would indicate that we have nothing to write, so we have successfully flushed everything.

File: src/micro_http/src/server.rs (hunk 25 above)
Comment (berciuliviu) [reply src/micro_http/src/server.rs:390]: From what I see, the connection state is also set to `AwaitingIncoming` on a successful write.

File: src/micro_http/src/server.rs (hunk 26 above)
Comment (berciuliviu) [reply src/micro_http/src/server.rs:392]: I agree with @georgepisaltu . Any other error is caught inside the write function. If we still want to log an error, the actual logging could be moved inside the write function itself. What do you think?

File: tests/integration_tests/functional/test_snapshot_basic.py [hunk 27]
```diff
@@ -352,6 +355,88 @@ def test_5_inc_snapshots(network_config,
...
+    3. Verify that an error was shown and the FC process is terminated.
+    """
+    logger = logging.getLogger("snapshot_load_failure")
+    vm = MicrovmBuilder(bin_cloner_path).create_basevm()
+    vm.spawn(log_level='Info')
+
+    vm.create_tap_and_ssh_config(host_ip=DEFAULT_HOST_IP,
+                                 guest_ip=DEFAULT_GUEST_IP,
+                                 netmask_len=DEFAULT_NETMASK,
+                                 tapname=DEFAULT_TAP_NAME)
```
Comment (acatangiu) [inline tests/integration_tests/functional/test_snapshot_basic.py:396]: nit: from what I can tell, you don't need a TAP or configuring microVM networking for this test
```suggestion
```

**Review by acatangiu — DISMISSED** [review]

File: tests/integration_tests/functional/test_snapshot_basic.py (hunk 27 above)
Comment (berciuliviu) [reply tests/integration_tests/functional/test_snapshot_basic.py:396]: You are correct

File: src/vmm/src/builder.rs [hunk 28]
```diff
@@ -419,23 +419,32 @@ pub fn build_microvm_from_snapshot(
             .map_err(RestoreMicrovmState)?;
 
     // Move vcpus to their own threads and start their state machine in the 'Paused' state.
-    vmm.start_vcpus(vcpus, seccomp_filter)
-        .map_err(StartMicrovmError::Internal)?;
+    vmm.start_vcpus(vcpus, seccomp_filter).map_err(|e| {
+        let _ = vmm.exit_vcpus();
```
Comment (acatangiu) [inline src/vmm/src/builder.rs:423]: I just realized that we can avoid carrying this around every error-case exit by moving it to `Vmm::drop()`.

That would also be a more complete fix; we always want to stop vcpus on dropping the vmm that owns them, right?

File: src/micro_http/src/server.rs (hunk 26 above)
Comment (berciuliviu) [reply src/micro_http/src/server.rs]: Done!

File: src/vmm/src/builder.rs (hunk 28 above)
Comment (berciuliviu) [reply src/vmm/src/builder.rs]: Correct, I have implemented the approach you provided!

**Review by acatangiu — APPROVED** [review]
LGTM!

**Review by sandreim — APPROVED** [review]
